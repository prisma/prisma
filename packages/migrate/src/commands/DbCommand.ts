import type { Command, Commands } from '@prisma/internals'
import { arg, format, HelpError, isError, unknownCommand } from '@prisma/internals'
import chalk from 'chalk'

export class DbCommand implements Command {
  public static new(cmds: Commands): DbCommand {
    return new DbCommand(cmds)
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('üèãÔ∏è  ')}Manage your database schema and lifecycle during development.

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db [command] [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
    --schema   Custom path to your Prisma schema

${chalk.bold('Commands')}
     pull   Pull the state from the database to the Prisma schema using introspection
     push   Push the state from Prisma schema to the database during prototyping
     seed   Seed your database
  execute   Execute native commands to your database

${chalk.bold('Examples')}

  Run \`prisma db pull\`
  ${chalk.dim('$')} prisma db pull

  Run \`prisma db push\`
  ${chalk.dim('$')} prisma db push

  Run \`prisma db seed\`
  ${chalk.dim('$')} prisma db seed

  Run \`prisma db execute\`
  ${chalk.dim('$')} prisma db execute
`)

  private constructor(private readonly cmds: Commands) {}

  /* eslint-disable-next-line @typescript-eslint/require-await */
  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--preview-feature': Boolean,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    // display help for help flag or no subcommand
    if (args._.length === 0 || args['--help']) {
      return this.help()
    }

    // check if we have that subcommand
    const cmd = this.cmds[args._[0]]
    if (cmd) {
      const argsForCmd = args['--preview-feature'] ? [...args._.slice(1), `--preview-feature`] : args._.slice(1)
      return cmd.parse(argsForCmd)
    }

    return unknownCommand(DbCommand.help, args._[0])
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbCommand.help}`)
    }
    return DbCommand.help
  }
}
