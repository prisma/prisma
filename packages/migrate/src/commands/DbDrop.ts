import {
  arg,
  canPrompt,
  checkUnsupportedDataProxy,
  Command,
  dropDatabase,
  format,
  getSchemaDir,
  HelpError,
  isError,
  link,
  loadEnvFile,
} from '@prisma/internals'
import chalk from 'chalk'
import prompt from 'prompts'

import { getDatasourceInfo } from '../utils/ensureDatabaseExists'
import { DbNeedsForceError } from '../utils/errors'
import { PreviewFlagError } from '../utils/flagErrors'
import { getSchemaPathAndPrint } from '../utils/getSchemaPathAndPrint'
import { printDatasource } from '../utils/printDatasource'

export class DbDrop implements Command {
  public static new(): DbDrop {
    return new DbDrop()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('ðŸ’£  ')}Drop the database

${chalk.bold.yellow('WARNING')} ${chalk.bold(
    `Prisma db drop is currently in Preview (${link('https://pris.ly/d/preview')}).
There may be bugs and it's not recommended to use it in production environments.`,
  )}
${chalk.dim('When using any of the subcommands below you need to explicitly opt-in via the --preview-feature flag.')}

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db drop [options] --preview-feature

${chalk.bold('Options')}

   -h, --help   Display this help message
     --schema   Custom path to your Prisma schema
  -f, --force   Skip the confirmation prompt

${chalk.bold('Examples')}

  Drop the database
  ${chalk.dim('$')} prisma db drop --preview-feature

  Specify a schema
  ${chalk.dim('$')} prisma db drop --preview-feature --schema=./schema.prisma

  Use --force to skip the confirmation prompt
  ${chalk.dim('$')} prisma db drop --preview-feature --force
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--preview-feature': Boolean,
      '--force': Boolean,
      '-f': '--force',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    await checkUnsupportedDataProxy('db drop', args, true)

    if (args['--help']) {
      return this.help()
    }

    if (!args['--preview-feature']) {
      throw new PreviewFlagError()
    }

    loadEnvFile(args['--schema'], true)

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const datasourceInfo = await getDatasourceInfo({ schemaPath, throwIfEnvError: true })
    printDatasource({ datasourceInfo })

    const schemaDir = (await getSchemaDir(schemaPath))!

    console.info() // empty line

    if (!args['--force']) {
      if (!canPrompt()) {
        throw new DbNeedsForceError('drop')
      }

      const confirmation = await prompt({
        type: 'text',
        name: 'value',
        message: `Enter the ${datasourceInfo.prettyProvider} database name "${
          datasourceInfo.dbName
        }" to drop it.\nLocation: "${datasourceInfo.dbLocation}".\n${chalk.red('All data will be lost')}.`,
      })
      console.info() // empty line

      if (!confirmation.value) {
        console.info('Drop cancelled.')
        // Return SIGINT exit code to signal that the process was cancelled.
        process.exit(130)
      } else if (confirmation.value !== datasourceInfo.dbName) {
        throw Error(`The database name entered "${confirmation.value}" doesn't match "${datasourceInfo.dbName}".`)
      }
    }

    // Url exists because we set `throwIfEnvErrors: true` in `getDatasourceInfo`
    if (await dropDatabase(datasourceInfo.url!, schemaDir)) {
      return `${process.platform === 'win32' ? '' : 'ðŸš€  '}The ${datasourceInfo.prettyProvider} database "${
        datasourceInfo.dbName
      }" from "${datasourceInfo.dbLocation}" was successfully dropped.\n`
    } else {
      return ''
    }
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbDrop.help}`)
    }
    return DbDrop.help
  }
}
