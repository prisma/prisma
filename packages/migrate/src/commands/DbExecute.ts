import {
  arg,
  checkUnsupportedDataProxy,
  Command,
  format,
  getCommandWithExecutor,
  getSchemaPath,
  HelpError,
  isError,
  loadEnvFile,
  logger,
} from '@prisma/internals'
import chalk from 'chalk'
import fs from 'fs'
import getStdin from 'get-stdin'
import path from 'path'

import { Migrate } from '../Migrate'
import type { EngineArgs } from '../types'

const helpOptions = format(
  `${chalk.bold('Usage')}

${chalk.dim('$')} prisma db execute [options]

${chalk.bold('Options')}

-h, --help            Display this help message

${chalk.italic('Datasource input, only 1 must be provided:')}
--url                 URL of the datasource to run the command on
--schema              Path to your Prisma schema file to take the datasource URL from

${chalk.italic('Script input, only 1 must be provided:')}
--file                Path to a file. The content will be sent as the script to be executed

${chalk.bold('Flags')}

--stdin              Use the terminal standard input as the script to be executed`,
)

export class DbExecute implements Command {
  public static new(): DbExecute {
    return new DbExecute()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('üìù ')}Execute native commands to your database

This command takes as input a datasource, using ${chalk.green(`--url`)} or ${chalk.green(
    `--schema`,
  )} and a script, using ${chalk.green(`--stdin`)} or ${chalk.green(`--file`)}.
The input parameters are mutually exclusive, only 1 of each (datasource & script) must be provided.
 
The output of the command is connector-specific, and is not meant for returning data, but only to report success or failure.

On SQL databases, this command takes as input a SQL script.
The whole script will be sent as a single command to the database.

${chalk.italic(`This command is currently not supported on MongoDB.`)}

${helpOptions}
${chalk.bold('Examples')}
 
  Execute the content of a SQL script file to the datasource URL taken from the schema
  ${chalk.dim('$')} prisma db execute
    --file ./script.sql \\
    --schema schema.prisma

  Execute the SQL script from stdin to the datasource URL specified via the \`DATABASE_URL\` environment variable
  ${chalk.dim('$')} echo 'TRUNCATE TABLE dev;' | \\
    prisma db execute \\
    --stdin \\
    --url="$DATABASE_URL"

  Like previous example, but exposing the datasource url credentials to your terminal history
  ${chalk.dim('$')} echo 'TRUNCATE TABLE dev;' | \\
    prisma db execute \\
    --stdin \\
    --url="mysql://root:root@localhost/mydb"
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--stdin': Boolean,
        '--file': String,
        '--schema': String,
        '--url': String,
        '--preview-feature': Boolean,
        '--telemetry-information': String,
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    await checkUnsupportedDataProxy('db execute', args, !args['--url'])

    if (args['--help']) {
      return this.help()
    }

    if (args['--preview-feature']) {
      logger.warn(`"prisma db execute" was in Preview and is now Generally Available.
You can now remove the ${chalk.red('--preview-feature')} flag.`)
    }

    loadEnvFile(args['--schema'], false)

    // One of --stdin or --file is required
    if (args['--stdin'] && args['--file']) {
      throw new Error(
        `--stdin and --file cannot be used at the same time. Only 1 must be provided. 
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    } else if (!args['--stdin'] && !args['--file']) {
      throw new Error(
        `Either --stdin or --file must be provided.
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    }

    // One of --url or --schema is required
    if (args['--url'] && args['--schema']) {
      throw new Error(
        `--url and --schema cannot be used at the same time. Only 1 must be provided.
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    } else if (!args['--url'] && !args['--schema']) {
      throw new Error(
        `Either --url or --schema must be provided.
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    }

    let script = ''
    // Read file
    if (args['--file']) {
      try {
        script = fs.readFileSync(path.resolve(args['--file']), 'utf-8')
      } catch (e) {
        if (e.code === 'ENOENT') {
          throw new Error(`Provided --file at ${args['--file']} doesn't exist.`)
        } else {
          console.error(`An error occurred while reading the provided --file at ${args['--file']}`)
          throw e
        }
      }
    }
    // Read stdin
    if (args['--stdin']) {
      script = await getStdin()
    }

    let datasourceType: EngineArgs.DbExecuteDatasourceType

    // Execute command(s) to url passed
    if (args['--url']) {
      datasourceType = {
        tag: 'url',
        url: args['--url'],
      }
    }
    // Execute command(s) to url from schema
    else {
      // validate that schema file exists
      // throws an error if it doesn't
      const schemaPath = await getSchemaPath(args['--schema'])

      // Execute command(s) to url from schema
      datasourceType = {
        tag: 'schema',
        // if schemaPath is undefined, getSchemaPath will error
        schema: schemaPath!,
      }
    }

    const migrate = new Migrate()

    try {
      await migrate.engine.dbExecute({
        script,
        datasourceType,
      })
    } finally {
      migrate.stop()
    }

    return `Script executed successfully.`
  }

  public help(error?: string): string | HelpError {
    if (error) {
      throw new HelpError(`\n${error}\n\n${helpOptions}`)
    }
    return DbExecute.help
  }
}
