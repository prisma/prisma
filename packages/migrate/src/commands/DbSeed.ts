import { arg, Command, format, getSchemaPath, HelpError, isError, loadEnvFile, logger } from '@prisma/internals'
import chalk from 'chalk'

import {
  executeSeedCommand,
  getSeedCommandFromPackageJson,
  legacyTsNodeScriptWarning,
  verifySeedConfigAndReturnMessage,
} from '../utils/seed'

export class DbSeed implements Command {
  public static new(): DbSeed {
    return new DbSeed()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('ðŸ™Œ  ')}Seed your database

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db seed [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--preview-feature': Boolean,
        '--schema': String,
        '--telemetry-information': String,
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (args['--preview-feature']) {
      logger.warn(`Prisma "db seed" was in Preview and is now Generally Available.
You can now remove the ${chalk.red('--preview-feature')} flag.`)

      // Print warning if user has a "ts-node" script in their package.json, not supported anymore
      await legacyTsNodeScriptWarning()
    }

    loadEnvFile(args['--schema'], true)

    // Print warning if user is using --schema
    if (args['--schema']) {
      logger.warn(
        chalk.yellow(
          `The "--schema" parameter is not used anymore by "prisma db seed" since version 3.0 and can now be removed.`,
        ),
      )
    }

    const seedCommandFromPkgJson = await getSeedCommandFromPackageJson(process.cwd())

    if (!seedCommandFromPkgJson) {
      // Only used to help users to set up their seeds from old way to new package.json config
      const schemaPath = await getSchemaPath(args['--schema'])

      const message = await verifySeedConfigAndReturnMessage(schemaPath)
      // Error because setup of the feature needs to be done
      if (message) {
        throw new Error(message)
      }

      return ``
    }

    // Seed command is set
    // Execute user seed command
    const successfulSeeding = await executeSeedCommand(seedCommandFromPkgJson)
    if (successfulSeeding) {
      return `\n${process.platform === 'win32' ? '' : 'ðŸŒ±  '}The seed command has been executed.`
    } else {
      process.exit(1)
    }
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbSeed.help}`)
    }
    return DbSeed.help
  }
}
