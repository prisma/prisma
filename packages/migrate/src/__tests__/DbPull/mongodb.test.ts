// describeIf is making eslint unhappy about the test names
/* eslint-disable jest/no-identical-title */

import { jestConsoleContext, jestContext } from '@prisma/get-platform'

import { DbPull } from '../../commands/DbPull'
import CaptureStdout from '../__helpers__/captureStdout'

const isMacOrWindowsCI = Boolean(process.env.CI) && ['darwin', 'win32'].includes(process.platform)
if (isMacOrWindowsCI) {
  jest.setTimeout(60_000)
}

const describeIf = (condition: boolean) => (condition ? describe : describe.skip)

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

// To avoid the loading spinner locally
process.env.CI = 'true'

describeIf(!process.env.TEST_SKIP_MONGODB)('MongoDB', () => {
  const captureStdout = new CaptureStdout()

  beforeEach(() => {
    captureStdout.startCapture()
  })

  afterEach(() => {
    captureStdout.clearCaptureText()
  })

  afterAll(() => {
    captureStdout.stopCapture()
  })

  const MONGO_URI = process.env.TEST_MONGO_URI_MIGRATE!

  if (isMacOrWindowsCI) {
    jest.setTimeout(60_000)
  }

  test('basic introspection', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--schema=./prisma/no-model.prisma'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)

    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "Prisma schema loaded from prisma/no-model.prisma

      Datasource "my_db": MongoDB database "tests-migrate" at "localhost:27017"



      - Introspecting based on datasource defined in prisma/no-model.prisma

      ✔ Introspected 1 model and 2 embedded documents and wrote them into prisma/no-model.prisma in XXXms
            
      *** WARNING ***

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Model: "users", field: "numberOrString1", original data type: "Json"

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
        - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"

      Run prisma generate to generate Prisma Client.
      "
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('introspection --force (existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--force'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)

    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "Prisma schema loaded from prisma/schema.prisma

      Datasource "my_db": MongoDB database "tests-migrate" at "localhost:27017"



      - Introspecting based on datasource defined in prisma/schema.prisma

      ✔ Introspected 1 model and 2 embedded documents and wrote them into prisma/schema.prisma in XXXms
            
      *** WARNING ***

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Model: "users", field: "numberOrString1", original data type: "Json"

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
        - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"

      Run prisma generate to generate Prisma Client.
      "
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('introspection --print (no existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--schema=./prisma/no-model.prisma', '--print'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)
    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "generator client {
        provider = "prisma-client-js"
      }

      datasource my_db {
        provider = "mongodb"
        url      = env("TEST_MONGO_URI_MIGRATE")
      }

      type UsersHobbies {
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString2 Json?
        objects         UsersHobbiesObjects[]
        tags            String[]
      }

      type UsersHobbiesObjects {
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString3 Json
        tags            String[]
      }

      model users {
        id              String         @id @default(auto()) @map("_id") @my_db.ObjectId
        admin           Boolean
        email           String
        hobbies         UsersHobbies[]
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString1 Json
      }

      "
    `)

    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      "
      // *** WARNING ***
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Model: "users", field: "numberOrString1", original data type: "Json"
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
      //   - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"
      // "
    `)
  })

  test('introspection --print --composite-type-depth=0 (no existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--schema=./prisma/no-model.prisma', '--print', '--composite-type-depth=0'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)
    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "generator client {
        provider = "prisma-client-js"
      }

      datasource my_db {
        provider = "mongodb"
        url      = env("TEST_MONGO_URI_MIGRATE")
      }

      model users {
        id              String  @id @default(auto()) @map("_id") @my_db.ObjectId
        admin           Boolean
        email           String
        hobbies         Json[]
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString1 Json
      }

      "
    `)

    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      "
      // *** WARNING ***
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Model: "users", field: "numberOrString1", original data type: "Json"
      // "
    `)
  })

  test('introspection --print --composite-type-depth=1 (no existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--schema=./prisma/no-model.prisma', '--print', '--composite-type-depth=1'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)
    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "generator client {
        provider = "prisma-client-js"
      }

      datasource my_db {
        provider = "mongodb"
        url      = env("TEST_MONGO_URI_MIGRATE")
      }

      type UsersHobbies {
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString2 Json?
        objects         Json[]
        tags            String[]
      }

      model users {
        id              String         @id @default(auto()) @map("_id") @my_db.ObjectId
        admin           Boolean
        email           String
        hobbies         UsersHobbies[]
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString1 Json
      }

      "
    `)

    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      "
      // *** WARNING ***
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Model: "users", field: "numberOrString1", original data type: "Json"
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
      // "
    `)
  })

  test('introspection --force --composite-type-depth=-1 (existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--force', '--composite-type-depth=-1'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)

    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "Prisma schema loaded from prisma/schema.prisma

      Datasource "my_db": MongoDB database "tests-migrate" at "localhost:27017"



      - Introspecting based on datasource defined in prisma/schema.prisma

      ✔ Introspected 1 model and 2 embedded documents and wrote them into prisma/schema.prisma in XXXms
            
      *** WARNING ***

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Model: "users", field: "numberOrString1", original data type: "Json"

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
        - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"

      Run prisma generate to generate Prisma Client.
      "
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('introspection --print --composite-type-depth=-1 (no existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--schema=./prisma/no-model.prisma', '--print', '--composite-type-depth=-1'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)
    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "generator client {
        provider = "prisma-client-js"
      }

      datasource my_db {
        provider = "mongodb"
        url      = env("TEST_MONGO_URI_MIGRATE")
      }

      type UsersHobbies {
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString2 Json?
        objects         UsersHobbiesObjects[]
        tags            String[]
      }

      type UsersHobbiesObjects {
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString3 Json
        tags            String[]
      }

      model users {
        id              String         @id @default(auto()) @map("_id") @my_db.ObjectId
        admin           Boolean
        email           String
        hobbies         UsersHobbies[]
        name            String
        /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
        numberOrString1 Json
      }

      "
    `)

    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      "
      // *** WARNING ***
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Model: "users", field: "numberOrString1", original data type: "Json"
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
      //   - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"
      // "
    `)
  })

  test('basic introspection --url', async () => {
    const introspect = new DbPull()
    const result = introspect.parse(['--print', '--url', MONGO_URI])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)

    expect(captureStdout.getCapturedText().join('\n')).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      "
      // *** WARNING ***
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Model: "users", field: "numberOrString1", original data type: "Json"
      // 
      // The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
      //   - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
      //   - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"
      // "
    `)
  })

  // In this case it should not error and the line `Datasource "x"` not be printed
  test('introspection --url - only generator defined', async () => {
    ctx.fixture('schema-only-mongodb/only-generator')
    const introspect = new DbPull()
    const result = introspect.parse(['--url', MONGO_URI])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)

    expect(captureStdout.getCapturedText().join('\n')).not.toContain(`Datasource `)
    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "Prisma schema loaded from schema.prisma



      - Introspecting

      ✔ Introspected 1 model and 2 embedded documents and wrote them into schema.prisma in XXXms
            
      *** WARNING ***

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Model: "users", field: "numberOrString1", original data type: "Json"

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
        - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"

      Run prisma generate to generate Prisma Client.
      "
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('introspection with --force', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse(['--force'])
    await expect(result).resolves.toMatchInlineSnapshot(`""`)

    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "Prisma schema loaded from prisma/schema.prisma

      Datasource "my_db": MongoDB database "tests-migrate" at "localhost:27017"



      - Introspecting based on datasource defined in prisma/schema.prisma

      ✔ Introspected 1 model and 2 embedded documents and wrote them into prisma/schema.prisma in XXXms
            
      *** WARNING ***

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Model: "users", field: "numberOrString1", original data type: "Json"

      The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
        - Composite type: "UsersHobbies", field: "numberOrString2", chosen data type: "Json"
        - Composite type: "UsersHobbiesObjects", field: "numberOrString3", chosen data type: "Json"

      Run prisma generate to generate Prisma Client.
      "
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })

  test('re-introspection should error (not supported) (existing models)', async () => {
    ctx.fixture('schema-only-mongodb')
    const introspect = new DbPull()
    const result = introspect.parse([])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
      "Iterating on one schema using re-introspection with db pull is currently not supported with MongoDB provider.
      You can explicitly ignore and override your current local schema file with prisma db pull --force
      Some information will be lost (relations, comments, mapped fields, @ignore...), follow https://github.com/prisma/prisma/issues/9585 for more info."
    `)

    expect(captureStdout.getCapturedText().join('\n')).toMatchInlineSnapshot(`
      "Prisma schema loaded from prisma/schema.prisma

      Datasource "my_db": MongoDB database "tests-migrate" at "localhost:27017"
      "
    `)
    expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
  })
})
