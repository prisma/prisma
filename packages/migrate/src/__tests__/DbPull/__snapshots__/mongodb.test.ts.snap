// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoDB basic introspection --url 2`] = `
"datasource db {
  provider = "mongodb"
  url      = "mongodb://localhost:27017/tests-migrate"
}

type UsersHobbies {
  name            String
  /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
  numberOrString2 Json?
  objects         UsersHobbiesObjects[]
  tags            String[]
}

type UsersHobbiesObjects {
  name            String
  /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
  numberOrString3 Json
  tags            String[]
}

model users {
  id              String         @id @default(auto()) @map("_id") @db.ObjectId
  admin           Boolean
  email           String
  hobbies         UsersHobbies[]
  name            String
  /// Multiple data types found: String: 50%, Int: 50% out of 2 sampled entries
  numberOrString1 Json
}

"
`;
